### PR TITLE
Strutils new sets

### PR DESCRIPTION
**Pull request by kintrix007**
Made it more consistent how the sets are defined, also changed them to use set unions wherever possible.

~~Added `Vowels` and `Consonants`, and functions called `isVowelAscii` and `isConsonantAscii`.~~
~~Added set `Punctuation` for punctuation characters alongside with `isPunctAscii` and `removePunctAscii`.~~
~~Added set `Special` for all non-alphanumerical characters with `isSpecialAscii` and `removeSpecialAscii` functions.~~

Added sets `ControlChars`, `GraphicChars`, `PrintableChars`, and `Punctuation`.

Just some simple additions that come in handy when dealing with text.

## future work
- [x] ControlChars, PrintableChars, GraphicChars (see comments)

### CODE REVIEW

File: lib/pure/strutils.nim [hunk 1]
```diff
@@ -265,8 +352,10 @@ func capitalizeAscii*(s: string): string {.rtl, extern: "nsuCapitalizeAscii".} =
   runnableExamples:
     doAssert capitalizeAscii("foo") == "Foo"
     doAssert capitalizeAscii("-bar") == "-bar"
-  if s.len == 0: result = ""
-  else: result = toUpperAscii(s[0]) & substr(s, 1)
+  if s.len == 0:
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: avoid style changes in unrelated code

File: lib/pure/strutils.nim [hunk 2]
```diff
@@ -84,31 +84,42 @@ from std/private/strimpl import cmpIgnoreStyleImpl, cmpIgnoreCaseImpl, startsWit
...
-  IdentStartChars* = {'a'..'z', 'A'..'Z', '_'}
+  IdentStartChars* = Letters + {'_'}
     ## The set of characters an identifier can start with.
 
-  Newlines* = {'\13', '\10'}
+  Newlines* = {'\r', '\n'}
     ## The set of characters a newline terminator can start with (carriage
     ## return, line feed).
 
+  Punctuation* = {'!', ',', '.', ':', ';', '?'}
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: is there an official source for that? can you add it as a comment in code?
(and do other languages like python, C++ etc implement it the same)

File: lib/pure/strutils.nim [hunk 3]
```diff
@@ -254,6 +313,34 @@ func toUpperAscii*(s: string): string {.rtl, extern: "nsuToUpperAsciiStr".} =
...
+    if not (s[i] notin Punctuation):
+      result[i] = s[i]
+
+func removeSpecialAscii*(s: string): string {.rtl, extern: "nsuRemoveSpecialAscii".} =
+  ## returns the a version of `s`, where all Special characters are
+  ## removed. This means all non-alphanumeric characters.
+  ## 
+  ## This works on ASCII characters only.
+  runnableExamples:
+    doAssert removeSpecialAscii("*nuzzle-wuzzles, uwu*") == "nuzzlewuzzles uwu"
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: `nim doc -r --lib:lib lib/pure/strutils.nim` fails for me

File: lib/pure/strutils.nim [hunk 4]
```diff
@@ -84,31 +84,42 @@ from std/private/strimpl import cmpIgnoreStyleImpl, cmpIgnoreCaseImpl, startsWit
...
+  IdentStartChars* = Letters + {'_'}
     ## The set of characters an identifier can start with.
 
-  Newlines* = {'\13', '\10'}
+  Newlines* = {'\r', '\n'}
     ## The set of characters a newline terminator can start with (carriage
     ## return, line feed).
 
+  Punctuation* = {'!', ',', '.', ':', ';', '?'}
+    ## The set of all punctuation characters. ('!', ',', '.', ':', ';', '?')
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: redundant, nim doc already shows the value of the const (check it)
(plus, these would've needed to be escaped with backticks)

File: lib/pure/strutils.nim [hunk 5]
```diff
@@ -119,6 +130,10 @@ const
     ##   doAssert "01234".find(invalid) == -1
     ##   doAssert "01A34".find(invalid) == 2
 
+  Special* = AllChars - Letters - Digits
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: not sure we need this category, it's subjective/not standard

File: lib/pure/strutils.nim [hunk 6]
```diff
@@ -84,31 +84,42 @@ from std/private/strimpl import cmpIgnoreStyleImpl, cmpIgnoreCaseImpl, startsWit
...
+  Whitespace* = {' ', '\t', '\v', '\r', '\n', '\f'}
     ## All the characters that count as whitespace (space, tab, vertical tab,
     ## carriage return, new line, form feed).
 
   Letters* = {'A'..'Z', 'a'..'z'}
     ## The set of letters.
 
+  Vowels* = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
+    ## The set of vowels.
+
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: in future work, I'd also like these sets:
* `ControlChars` (with semantics derived from `iscntrl`)
* `PrintableChars` (with semantics derived from `isprint`)

File: lib/pure/strutils.nim [hunk 7]
```diff
@@ -84,31 +84,42 @@ from std/private/strimpl import cmpIgnoreStyleImpl, cmpIgnoreCaseImpl, startsWit
 
 
 const
-  Whitespace* = {' ', '\t', '\v', '\r', '\l', '\f'}
+  Whitespace* = {' ', '\t', '\v', '\r', '\n', '\f'}
```
Comment (timotheecour) [inline lib/pure/strutils.nim:87]: also, a separate PR would be welcome to change all instances of `\l` to `\n` in nim repo (and it'd only do that plus maybe related changes)

File: lib/pure/strutils.nim [hunk 8]
```diff
@@ -130,6 +145,24 @@ func isAlphaAscii*(c: char): bool {.rtl, extern: "nsuIsAlphaAsciiChar".} =
     doAssert isAlphaAscii('8') == false
   return c in Letters
 
+func isVowelAscii*(c: char): bool {.rtl, extern: "nsuIsVowelAsciiChar".} =
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: what's the point? `c in Vowels` is good enough

ditto for isConsonantAscii
ditto for isPunctAscii
ditto for isSpecialAscii

File: lib/pure/strutils.nim [hunk 9]
```diff
@@ -254,6 +313,34 @@ func toUpperAscii*(s: string): string {.rtl, extern: "nsuToUpperAsciiStr".} =
     doAssert toUpperAscii("FooBar!") == "FOOBAR!"
   toImpl toUpperAscii
 
+func removePunctAscii*(s: string): string {.rtl, extern: "nsuRemovePunctAscii".} =
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: this is bloat; remove both removePunctAscii and removeSpecialAscii, they're best replaced by a more generic and flexible (and efficient!) alternative that would filter in-place a string based on a set (if it doesn't exist we can add this in a separate PR to std/strbasics)

File: lib/pure/strutils.nim (hunk 1 above)
Comment (kintrix007) [reply lib/pure/strutils.nim]: Ah, sorry, it must have been a reflex I changed that... It was next to some code I changed...
Will make sure not to do that! Thanks

File: lib/pure/strutils.nim (hunk 8 above)
Comment (kintrix007) [reply lib/pure/strutils.nim]: I just added them because functions for all other, like `isAlphaAscii` exist.
It would 100% make it easier if they didn't exist, I simply added them to make it more consistent with some of the other functions like this.

File: lib/pure/strutils.nim (hunk 6 above)
Comment (kintrix007) [reply lib/pure/strutils.nim]: I will definitely look into that!

File: lib/pure/strutils.nim (hunk 8 above)
Comment (timotheecour) [reply lib/pure/strutils.nim]: I think `isAlphaAscii` + friends only exist for historical reasons, consistency is less important than avoiding bloat/redundancy here

File: lib/pure/strutils.nim (hunk 5 above)
Comment (kintrix007) [reply lib/pure/strutils.nim]: Alright!

File: lib/pure/strutils.nim (hunk 2 above)
Comment (kintrix007) [reply lib/pure/strutils.nim]: This is just a set I made up, without a standard... which is a terrible idea now that you mention it...
Haskell does have an `isPunctuation` function, however, that works properly on all unicode characters.
I will change it to have all ASCII characters that are categorized as punctuation in Haskell.

Comment (timotheecour) [reply lib/pure/strutils.nim]: python3 -c 'import string; print(string.punctuation)'
!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~

please also do some research for other popular languages; if there's consensus among those, then adding it is a no-brainer, if not, then we need to think a bit

File: lib/pure/strutils.nim (hunk 6 above)
Comment (kintrix007) [reply lib/pure/strutils.nim]: I'm not sure how I should do that, and what exactly I should do... so I could probably do that in future PR.

File: lib/pure/strutils.nim (hunk 3 above)
Comment (kintrix007) [reply lib/pure/strutils.nim]: I initually added that as a joke, but it's strange it failed... :/
Well, it's getting removed, either way

File: lib/pure/strutils.nim (hunk 2 above)
Comment (kintrix007) [reply lib/pure/strutils.nim]: Yeah, Haskell lists characters a little differently...
I changed it to that temporarily, will look up some other languages

File: lib/pure/strutils.nim (hunk 6 above)
Comment (timotheecour) [reply lib/pure/strutils.nim]: in that future PR, you could iterate over all chars, and call iscntrl, isprint on it, and populate a set with the values from it:
```nim
const ControlChars = {'\12', ...}
  # computed it offline and verify it in tests via tests/stdlib/tstrutils.nim
```

```nim
# tstrutils.nim
proc iscntrl(c: cint): cint {.importc, header: "<ctype.h>".}
block: # ControlChars
  for i in char.low .. char.high:
    doAssert (iscntrl(i.cint) > 0) == (i in ControlChars)
```

File: lib/pure/strutils.nim (hunk 2 above)
Comment (timotheecour) [reply lib/pure/strutils.nim]: ~~looks like C and python are off by 1 char:~~

C and python agree (i was missing the `\` as 
python3 -c 'import string; print(string.punctuation)'` was interpretting \ on the shell so didn't show up in output)

=> so the correct definition should be:
```nim
const Punctuation = {'!', '\"', '#', '$', '%', '&', '\'', '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~'}
```

C and python in agreement seems like good enough of a standard

```nim
proc ispunct(a: cint): cint {.importc, header: "<ctype.h>".}
proc main()=
  var s: set[char]
  for c in char.low..char.high:
    if c.cint.ispunct > 0:
      s.incl c
  echo s
  var s2: set[char]
  for c in """!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~""":
    s2.incl c
  echo s2
  echo s == s2
main()
```

```
{'!', '\"', '#', '$', '%', '&', '\'', '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~'}
{'!', '\"', '#', '$', '%', '&', '\'', '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~'}
true
```

File: lib/pure/strutils.nim [hunk 10]
```diff
@@ -185,6 +193,10 @@ func isUpperAscii*(c: char): bool {.rtl, extern: "nsuIsUpperAsciiChar".} =
     doAssert isUpperAscii('7') == false
   return c in {'A'..'Z'}
 
+template toImpl(call) =
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: this can be moved back to where it was to reduce the diff

File: lib/pure/strutils.nim [hunk 11]
```diff
@@ -84,31 +84,39 @@ from std/private/strimpl import cmpIgnoreStyleImpl, cmpIgnoreCaseImpl, startsWit
...
-    ## return, line feed).
+  Newlines* = {'\r', '\n'}
+    ## The set of characters a newline terminator can consist of
+
+  Punctuation* = {'!'..'/'} - {'+', '$'} + {':', ';', '?', '@', '['..']', '_', '{', '}', '\161','\167','\171','\182','\183','\187','\191'}
+    ## The set of all ASCII punctuation characters.
 
   AllChars* = {'\x00'..'\xFF'}
-    ## A set with all the possible characters.
+    ## The set with all the possible characters.
```
Comment (timotheecour) [inline lib/pure/strutils.nim]:
```suggestion
    ## The set of all characters.
```

File: lib/pure/strutils.nim [hunk 12]
```diff
@@ -84,31 +84,39 @@ from std/private/strimpl import cmpIgnoreStyleImpl, cmpIgnoreCaseImpl, startsWit
...
     ## All the characters that count as whitespace (space, tab, vertical tab,
     ## carriage return, new line, form feed).
 
   Letters* = {'A'..'Z', 'a'..'z'}
     ## The set of letters.
 
+  Vowels* = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
+    ## The set of vowels.
+  
+  Consonants* = Letters - Vowels
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: update changelog.md: Added xyz

File: lib/pure/strutils.nim [hunk 13]
```diff
@@ -84,31 +84,39 @@ from std/private/strimpl import cmpIgnoreStyleImpl, cmpIgnoreCaseImpl, startsWit
...
+  IdentStartChars* = Letters + {'_'}
     ## The set of characters an identifier can start with.
 
-  Newlines* = {'\13', '\10'}
-    ## The set of characters a newline terminator can start with (carriage
-    ## return, line feed).
+  Newlines* = {'\r', '\n'}
+    ## The set of characters a newline terminator can consist of
+
+  Punctuation* = {'!'..'/'} - {'+', '$'} + {':', ';', '?', '@', '['..']', '_', '{', '}', '\161','\167','\171','\182','\183','\187','\191'}
```
Comment (timotheecour) [inline lib/pure/strutils.nim]:
```nim
  Punctuation = {'!', '\"', '#', '$', '%', '&', '\'', '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~'}
```

refs https://github.com/nim-lang/Nim/pull/18193#discussion_r646047528

File: lib/pure/strutils.nim [hunk 14]
```diff
@@ -84,31 +84,41 @@ from std/private/strimpl import cmpIgnoreStyleImpl, cmpIgnoreCaseImpl, startsWit
...
 
-  Newlines* = {'\13', '\10'}
-    ## The set of characters a newline terminator can start with (carriage
-    ## return, line feed).
+  Newlines* = {'\r', '\n'}
+    ## The set of characters a newline terminator can consist of
+
+  # same as {'!', '\"', '#', '$', '%', '&', '\'', '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~'}
+  # same as {'!'..'/', ':'..'@', '['..'`', '{'..'~'}
+  Punctuation* = {'!'..'~'} - Letters - Digits
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: why not give the list so it's self-documenting?

```
  Punctuation* = {'!', '\"', '#', '$', '%', '&', '\'', '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~'}
```

(if you're worried about performance, it'd be the same)

File: lib/pure/strutils.nim (hunk 2 above)
Comment (Clyybber) [reply lib/pure/strutils.nim]: https://github.com/nim-lang/Nim/pull/16708

File: lib/pure/strutils.nim [hunk 15]
```diff
@@ -84,31 +84,41 @@ from std/private/strimpl import cmpIgnoreStyleImpl, cmpIgnoreCaseImpl, startsWit
...
 const
-  Whitespace* = {' ', '\t', '\v', '\r', '\l', '\f'}
+  Whitespace* = {' ', '\t', '\v', '\r', '\n', '\f'}
     ## All the characters that count as whitespace (space, tab, vertical tab,
     ## carriage return, new line, form feed).
 
   Letters* = {'A'..'Z', 'a'..'z'}
     ## The set of letters.
 
+  Vowels* = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
```
Comment (Clyybber) [inline lib/pure/strutils.nim]: This is not "standard" enough to be added IMO, for example `y` or `w` are considered vowels too in some languages and not in others.

File: lib/pure/strutils.nim (hunk 12 above)
Comment (Clyybber) [reply lib/pure/strutils.nim]: Same as https://github.com/nim-lang/Nim/pull/18193/files#r646103498 ; `r`,`y`,`w`,`h`

**Comment by juancarlospaco** [comment]
Should new additions go into `strbasics` ?.

File: lib/pure/strutils.nim (hunk 2 above)
Comment (timotheecour) [reply lib/pure/strutils.nim]: The comment that closed the other PR (https://github.com/nim-lang/Nim/pull/16708#issuecomment-760108509) would be justified if there was no definition or agreement on punctuation, but the fact that both C and python agree (assuming default c locale) is a strong argument.

There's actually a definition of punctuation:
> The standard "C" locale considers punctuation characters all graphic characters (as in isgraph) that are not alphanumeric (as in isalnum).

Applications can always choose to use a different set for their needs, but having an accepted standard makes sense.

File: lib/pure/strutils.nim (hunk 15 above)
Comment (timotheecour) [reply lib/pure/strutils.nim]: sadly i have to agree with @Clyybber; we should stick to C standard in strutils, refs https://www.tutorialspoint.com/c_standard_library/ctype_h.htm, which includes ispunct (this PR's Punctuation), and in future PR, isprint (PrintableChars), isgraph (GraphChars), iscntrl (ControlChars)

File: lib/pure/strutils.nim (hunk 12 above)
Comment (timotheecour) [reply lib/pure/strutils.nim]: not sure what you mean by rywh (granted, you could say same about xyz, but i was referring to fact that symbol additions should be added to changelog)

and yes, i agree that Vowels and Consonants can't be added because of y and w

File: lib/pure/strutils.nim (hunk 14 above)
Comment (kintrix007) [reply lib/pure/strutils.nim]: Yeah, I know that the performance is the same... Since, as far as I know, the set differences would be computed compile time, so those two are literally the same.
Gonna do another commit, and change it to that, but I was just unsure which one makes it more obvious. Since the version I left there shows that that means all characters except alphanumeric.
Changing it to `GraphicChars - Letters - Digits` might make more sense, when it's added

File: lib/pure/strutils.nim (hunk 12 above)
Comment (kintrix007) [reply lib/pure/strutils.nim]: Yep, removing them.

File: tests/stdlib/tstrutils.nim [hunk 16]
```diff
@@ -13,6 +13,33 @@ template rejectParse(e) =
...
+    proc isspace(c: cint): cint {.importc, header: "<ctype.h>".}
+    for i in char.low .. char.high:
+      doAssert (isspace(i.cint) > 0) == (i in Whitespace)
+
+  block: # ControlChars
+    proc iscntrl(c: cint): cint {.importc, header: "<ctype.h>".}
+    for i in char.low .. char.high:
+      doAssert (iscntrl(i.cint) > 0) == (i in ControlChars)
+  
+  block: # GraphicChars
```
Comment (kintrix007) [inline tests/stdlib/tstrutils.nim]: I managed to mess it up so that they do not even compile (^^;)
(I'm really new to Nim, and haven't ever used c imports)

File: lib/pure/strutils.nim [hunk 17]
```diff
@@ -112,10 +106,18 @@ const
...
+
+  GraphicChars* = {'!'..'~'}
+    ## The set of all ASCII graphic characters.
+  
+  PrintableChars* = GraphicChars + {' '}
+    ## The set of all ASCII printable characters.
+
+  Punctuation* = {'!', '\"', '#', '$', '%', '&', '\'', '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~'}
     ## The set of all ASCII punctuation characters.
+    ## Is the same as `PrintableChars - Letters - Digits`
```
Comment (timotheecour) [inline lib/pure/strutils.nim]:
```suggestion
    ## Is the same as `GraphicChars - Letters - Digits`
```

(seemed like a bug! also, add this as a test in tstrutils)

File: lib/pure/strutils.nim [hunk 18]
```diff
@@ -112,10 +106,18 @@ const
   Newlines* = {'\r', '\n'}
     ## The set of characters a newline terminator can consist of
 
-  # same as {'!', '\"', '#', '$', '%', '&', '\'', '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~'}
-  # same as {'!'..'/', ':'..'@', '['..'`', '{'..'~'}
-  Punctuation* = {'!'..'~'} - Letters - Digits
+  ControlChars* = {'\0'..'\31', '\127'}
+    ## The set of all ASCII control characters.
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: how about:
```nim
    ## The set of all ASCII control characters, i.e. that does not occupy
    ## a printing position on a display
    ## (see also https://www.cplusplus.com/reference/cctype/iscntrl/)
```

likewise, PrintableChars doc comment could similarly be expanded, eg:
```nim
    ## The set of all ASCII printable characters, i.e. that occupy a printing
    ## position on a display (see also https://www.cplusplus.com/reference/cctype/isprint/)
```

the rest are clear as they can be derived from other definitions

**Comment by timotheecour** [comment]
for CI failures, try:
```nim
when nimvm: discard
else:
  ... # things with importc
```

(but test locally before pushing ;-) )

**Review by kintrix007 — COMMENTED** [review]
> for CI failures, try:
> 
> ```nim
> when nimvm: discard
> else:
>   ... # things with importc
> ```
> 
> (but test locally before pushing ;-) )

I did test locally, but I had no idea how to fix it, so I pushed it to get help, and I got help :P

File: lib/pure/strutils.nim [hunk 19]
```diff
@@ -94,21 +94,38 @@ const
...
+
+  ControlChars* = {'\0'..'\31', '\127'}
+    ## The set of all ASCII control characters, i.e. that do not occupy
+    ## a printing position on a display.
+    ## (see also https://www.cplusplus.com/reference/cctype/iscntrl/)
+
+  GraphicChars* = {'!'..'~'}
+    ## The set of all ASCII graphic characters, i.e. that occupy a printing
+    ## position on a display and have a graphical representation.
+    ## (see also https://www.cplusplus.com/reference/cctype/isprint/)
```
Comment (timotheecour) [inline lib/pure/strutils.nim]: wrong link

**Review by timotheecour — DISMISSED** [review]
LGTM after addressing https://github.com/nim-lang/Nim/pull/18193/files#r646179344

**Comment by kintrix007** [comment]
Oof, most tests fail with JS... If I understand it correctly, they fail on the set checks, where we check if they align with the C stdlib...

File: lib/pure/strutils.nim (hunk 2 above)
Comment (Clyybber) [reply lib/pure/strutils.nim]: > The standard "C" locale considers punctuation characters all graphic characters (as in isgraph) that are not alphanumeric (as in isalnum).

Then there is no need to add a `Punctuation` set that someone will have to look up.how it is specified because `Printable - AlphaNumeric` is much clearer and "self-documenting".

**Comment by timotheecour** [comment]
> Oof, most tests fail with JS... If I understand it correctly, they fail on the set checks, where we check if they align with the C stdlib...

this works:
```nim
from stdtest/testutils import disableVm
=>
from stdtest/testutils import disableVm, whenVMorJs

```
and then:
```nim
  whenVMorJs: discard
  do:
    block: # Whitespace
      proc isspace(c: cint): cint {.importc, header: "<ctype.h>".}
    ...
```

(EDIT: i pushed it in your branch)

**Comment by Araq** [comment]
The refactorings are fine and would be accepted but I consider `GraphicChars` and others to be a misdesign: 

- It's clear that these are not universal definitions so people coming from C disagree with people coming from Python. Or maybe an upcoming C23 standard changes these definitions because of locals or unicode, it's hitting a moving target.
- So then we have our own defintion what `GraphicChars` means and we cannot change it ever without breaking code. Suppose I am to implement spec X. Spec X also uses a notion of GraphicChars and defines what it means. Is it the same definition as Nim's? If so, that would be accidentical so I better write my own set definition anyway. If it differs slightly I have no chance anyway but to write my own definition. Either way we're better off not offering this at all, as the good code doesn't use it anyway. We lure people into the wrong solution by making strutils.GraphicGhars available. Much like we encourge bugs by `strutils.escape`'s very existence.

**Comment by timotheecour** [comment]
> It's clear that these are not universal definitions so people coming from C disagree with people coming from Python

python and C actually agree on all those definitions, as noted in review comments. Vowels and Consonants (in a previous version of this PR) have been removed because of lack of standardized definition.

this PR follows a widely accepted standard, the posix character class, and lots of languages follow this standard:

* D: https://dlang.org/phobos/std_ascii.html (see implementation, i checked it gives same results)
* python3: see https://docs.python.org/3/library/string.html (printable, punctuation etc) and https://docs.python.org/3/library/curses.ascii.html (isgraph, ispunct etc)
* C: `ctype.h`:  http://www.cplusplus.com/reference/cctype/
* C++: `<cctype>`: std::ispunct etc https://en.cppreference.com/w/cpp/string/byte/ispunct
* regular expression: see `[:punct:]` from https://en.wikipedia.org/wiki/Regular_expression#POSIX_character_classes, and also, `[:graph:]` , `[:print:]`, `[:cntrl:]` and the rest

etc

> Or maybe an upcoming C23 standard changes these definitions because of locals or unicode, it's hitting a moving target.

they won't change those definitions, that'd break code; if needed they'll simply add another character class

> So then we have our own defintion what GraphicChars means and we cannot change it ever without breaking code.

ditto, we'll just add a new character class if ever needed

ascii is stable and has been around for a while,  it's not a moving target. Nim should follow widely accepted standards and not make client code re-invent the wheel (inconsistently)

**Comment by Araq** [comment]
> they won't change those definitions, that'd break code; if needed they'll simply add another character class

Ok, so then we "only" need to add more of these classes for consistency with Posix. IMHO these "character classes" exist because the respective environments lack Nim's set construct. The classes feel archaic in the days of Unicode. Who cares if the character is an 
 ASCII control character -- what are the use cases? "Other languages copied anachronisms from each other" is not a use case.

**Comment by timotheecour** [comment]
input validation/filtering is a typical use case.
I agree that `std/strutils` is bloated though so I'd be ok with either moving those new sets to `std/strmisc` (2nd choice) or adding `std/ascii` (1st choice, can grow in future if needed yet avoids scope creep as it's right there in the title)

D and python3 also have such an ascii modules (https://dlang.org/phobos/std_ascii.html, https://docs.python.org/3/library/curses.ascii.html)

**Comment by Araq** [comment]
> I agree that std/strutils is bloated though so I'd be ok with either moving those new sets to std/strmisc (2nd choice) or adding std/ascii (1st choice, can grow in future if needed yet avoids scope creep as it's right there in the title)

It should be `dist/ascii`. We should use `std` for std-worthy stuff. Yes, I know this is a change in policy.

**Comment by timotheecour** [comment]
> It should be dist/ascii.

you mean `stdx/ascii` (in $nim/lib/stdx/ascii.nim) ? if so I'm all for it

refs: https://github.com/nim-lang/Nim/pull/17722#issuecomment-820659976
> Ok, let's use "stdx" then.

(separate topic but stdx/chains instead of std/chains would be nice too)

**Comment by Araq** [comment]
> you mean stdx/ascii (in $nim/lib/stdx/ascii.nim) ? if so I'm all for it

No, I mean `dist/ascii`. Or some similar name that implies "it's simply a module we ship with Nim".  `stdx` still has "standard" in its name implying a level of rigor that we lack.

**Comment by timotheecour** [comment]
not sure what we gain by multiplying the number of top-level prefixes:
```nim
import std/strbasics
import experimental/diff
import packages/docutils/rstgen
import stdx/syntaxes # still being debated, see also `extensions` (https://github.com/nim-lang/Nim/pull/17722#issuecomment-820601135)
import dist/ascii
```
not to mention all the ones where std prefix is optional, eg:
`import strutils`

A module-level metadata, top-level doc comment, or separate index would all be better suited than categorizing modules based on a subjective notion of fitness. After all we also already have `std/chains` and other modules which don't quite fit the bill.

That just invites parallel sub-categories, eg:
```nim
std/js/foo
dist/js/foo
```
which is a bad thing, for the same reason parallel APIs are bad.

Most other language's standard libraries manage to do with a single top-level prefix which avoids scope pollution.

**Comment by Araq** [comment]
> Most other language's standard libraries manage to do with a single top-level prefix which avoids scope pollution.

> https://docs.python.org/3/library/curses.ascii.html

But ok, we can also develop the standard library with rigor instead. "Here, I found X to be useful so I added it and C++ and Python also have it" is pretty far away from what I consider to be acceptable. Especially since these languages all lack Nim's set construct. 

And it's even worse, I cannot bring up the argument that "we don't need std/asciitables because Python also lacks std/asciitables" as you simply don't care when it's the other way round. Everything should always be added to "std", immediately, giving us the superset of things that are in C++ and Python **plus** all the things that happen to be useful inside the Nim compiler. It's terrible.

**Comment by timotheecour** [comment]
For the sake of moving forward with this, I'm ok with creating `lib/dist/` and adding `lib/dist/ascii.nim` so that modules needing it can import it as `import dist/ascii` (although I would've preferred `import std/ascii` or `import stdx/ascii`)

> std/asciitables

right now it's in `lib/std/private/asciitables.nim`, which means it's internal and client code isn't supposed to use it unless they know what they're doing and are ok with breaking changes, but I happen to find `asciitables` actually quite useful and std-worthy for future work, I often use it in my own code. Whether it shall be moved to std/asciitables, stdx/asciitables or dist/asciitables is rather secondary so long it's somewhere accessible under `lib` (honored by --lib flag)

so, ok to move forward with this PR after modifying it so that new sets go to `lib/dist/ascii.nim` and also keeping the other improvements from this PR? (the new tests should be moved, maybe under tests/stdlib/tascii for simplicity, instead of tests/dist/tascii, but I don't care too much which of those 2 files is used)

**Comment by Araq** [comment]
Let's go with `import dist/asciisets`

**Comment by timotheecour** [comment]
ok, ping @kintrix007 :)

(EDIT: ie, introduce `lib/dist` and `lib/dist/asciisets.nim`)

**Comment by kintrix007** [comment]
> ok, ping @kintrix007 :)
> 
> (EDIT: ie, introduce `lib/dist` and `lib/dist/asciisets.nim`)

First of all, sorry for being away for "a short while"...
But I'm kind of confused what we want to do now.

Do we want to move all string sets to `lib/dist/asciisets.nim`? If so, that would cause some back-wards incompatibility.
However, if we want to only move the newly created sets, then wouldn't it be awkward to have to import both `dist/asciisets` and `std/strutils` to have all sets to work with? As `dist/asciisets` would only have the sets like `PrintableChars` and so

**Comment by timotheecour** [comment]
> Do we want to move all string sets to lib/dist/asciisets.nim? If so, that would cause some back-wards incompatibility. However, if we want to only move the newly created sets, then wouldn't it be awkward to have to import both dist/asciisets and std/strutils to have all sets to work with? As dist/asciisets would only have the sets like PrintableChars and so

good that you raise this point.

I think the cleanest solution is as follows:
* create `lib/std/private/constants.nim` and add all the sets there
* make std/strutils import it and re-export the ones it already defined
* make dist/asciisets import it and re-export the ones that make sense there
* other modules that also copy-paste the definition of a few of those sets can now import/re-export those instead

I've been wanting such a module (`lib/std/private/constants.nim`) for other reasons (including some things shared between compiler and stdlib, refs https://github.com/timotheecour/Nim/issues/616), to refactor some duplicate definitions, so it could be reused for that purpose too. Note that `lib/std/private/constants.nim` is an implementation detail (it's right there in the name `private`) so it's fine to do so. 

The scope of `lib/std/private/constants.nim` could be restricted further if needed, but "useful constants" is good enough for now; again it's an implementation detail

## benefits
* single source of truth / documentation
* `lib/std/private/constants.nim` is cheap to semcheck, it's only constants and these are not code-gen'd
* no duplicate identifier error (it's the same symbol), no backward compatibility issue, no need to rename the sets to avoid clashes (renaming is bad IMO)
* only question is what docs would look like, but I think it's either fine, or docgen can be taught to handle re-exports in whatever way we want

more generally, I find the addition of lib/std/private truly liberating, much better than the convoluted ways that preceded it with system / include files

@araq wdyt ?

**Comment by kintrix007** [comment]
>     * make dist/asciisets import it and re-export the ones that make sense there

To me it would make sense to add all sets there. Since if it does not export all, then there is gonna be that awkwardness of having to also implement `strutils` just for a set it's missing. However, If we define all of the sets there as well, then if we import both `strutils` and `asciisets` then some of it would be double-declared... And so would force you to write `strutils.Letters` or `asciisets.Letters` while really they are the same.

I am not misunderstanding something, right?

**Comment by timotheecour** [comment]
>  if we import both strutils and asciisets then some of it would be double-declared... And so would force you to write strutils.Letters or asciisets.Letters while really they are the same.

there is no double-declaration if it's the same symbol (eg via import + export).

```nim
# main
import t12425b, t12425c, t12425d
echo A1

# t12425b
const A1* = {1,2,3} ## some comment

# t12425c
import t12425b
export A1

# t12425d
import t12425b
export A1
```
`nim r main` works

**Comment by timotheecour** [comment]
ping @kintrix007

**Comment by kintrix007** [comment]
I'm sorry, I do not really have time to do this around now, but will get to it when I have time and will.

**Comment by kintrix007** [comment]
I will not come back to this, sorry. If this is still needed, which I kind of doubt, someone else is probably gonna do it.
